Add trace function

dkhofer opened the pull request:
Hi,

I hope you don't mind unsolicited pull requests.  I pretty much just grabbed this from the main Crystal Matrix class: https://github.com/manastech/crystal/blob/master/src/matrix.cr#L691

I didn't immediately see a way to compute the trace using lapack, and since it's just adding n numbers together (for an n x n matrix) I think performance probably isn't a big deal.  But let me know if you have any comments or questions.  Thanks!

asterite commented on src/crystalla/matrix.cr:
```diff
@@ -341,6 +341,14 @@ module Crystalla
+    # Returns the sum of the diagonal elements. Only useful for numeric matrices.
+    def trace
+      raise ArgumentError.new "Number of rows (#{number_of_rows}) does not match number of columns (#{number_of_cols})" unless square?
+      (0...number_of_cols).inject(0) do |memo, i|
```
I'd use `(0...number_of_cols).sum(0.0) { |i| self[i, i] }`. Also please use `0.0` as the starting number, otherwise the result ends up being `Int32 | Float64`

asterite commented on spec/matrix_spec.cr:
```diff
@@ -403,6 +403,25 @@ describe Matrix do
   end
 
+  context "trace" do
+    it "computes trace for integers" do
```
I don't think this spec adds much value, a matrix is always backed by an array of floats

mverzilli commented:
Hey, thanks a lot! I didn't expect PRs this soon, but they're certainly welcome! I hope you bear with us for API changes, this is in a "Big Bang" stage so we'll be breaking everything with almost every single commit. 

I agree that it's probably overkill to use LAPACK or BLAS for this. What is more, @spalladino benchmarked a pure Crystal implementation of `sum` against BLAS and Crystal won. So for these kinds of really simple algorithms just using Crystal may be the right way to go.

I have only two things to ask before merging: 
1. Try to apply @asterite's suggestions.
2. Add a result type annotation to `trace`. We're going to start doing that with public APIs soon (maybe tonight!), because this is a lib and it will lead to more predictable behaviour across clients. 

Thanks a lot for your contribution!

asterite commented:
Also please squash your commits :-)

dkhofer committed:
Add trace function.

dkhofer commented:
Thanks for the feedback!  I believe I've addressed all comments.  

I didn't know about sum(), that's pretty cool.

I assume you wanted Float64 as a return type?

I had actually never squashed commits before, let me know if I did it wrong.  :-)
